src/west/commands/project.py

```diff
@@ -157,6 +159,86 @@ def do_run(self, args, user_args):
             log.inf(result)
 
 
+class ManifestCommand(WestCommand):
+    # The slightly weird naming is to avoid a conflict with
+    # west.manifest.Manifest.
+
+    def __init__(self):
+        super(ManifestCommand, self).__init__(
+            'manifest',
+            'manage the west manifest',
+            _wrap('''
+            Manages the west manifest.
+
+            Currently, only one operation on the manifest is
+            implemented, namely --freeze. This outputs the manifest
+            with all project-related values fully specified: defaults
+            are applied, and all revisions are converted to SHAs based
+            on the current manifest-rev branches.'''),
+            accepts_unknown_args=False)
+
+    def do_add_parser(self, parser_adder):
+        parser = parser_adder.add_parser(
+            self.name,
+            help=self.help,
+            description=self.description)
+
+        group = parser.add_mutually_exclusive_group(required=True)
+        group.add_argument('--freeze', action='store_true',
+                           help='emit a manifest with SHAs for each revision')
+
+        group = parser.add_argument_group('options for --freeze')
+        group.add_argument('-o', '--out',
+                           help='output file, default is standard output')
+
+        return parser
+
+    def do_run(self, args, user_args):
+        # We assume --freeze here. Future extensions to the group
+        # --freeze is part of can move this code around.
+
+        m = Manifest.from_file()
+
+        # Build a 'frozen' representation of all projects, except the
+        # manifest project.
+        projects = list(m.projects)
+        del projects[MANIFEST_PROJECT_INDEX]
+        frozen_projects = []
+        for project in projects:
+            sha = _sha(project, '{qual_manifest_rev_branch}')
```

Comment:
> It appears we are having an issue with the `_sha` method when referring to a PR.

This is a problem with the way we manage our ref space actually that probably requires more discussion between us. In my first try, this line:

```
            sha = _sha(project, '{qual_manifest_rev_branch}')
```

Was this:

```
            sha = _sha(project, project.revision)
```

But the problem was that this didn't work if the revision was a branch (which is exactly what it *must* do and the main reason for its existence is), because the current fetching code doesn't pull in the remote's head ref space (into `refs/west/*`) unless the revision is a SHA! :laughing: 

So my first response was to change the fetching code to just unconditionally always fetch the entire remote project ref space and try to freeze using that. But I thought this would be a more controversial change and that I could just get away with using manifest-rev for now. I see now that I was wrong, because west does not have a manifest-rev branch.

So our two options are:

1. Add a manifest-rev branch to west and use that (I actually kind of like that for consistency)
2. Fetch the entire remote project's ref space always in all cases and use that (I think this has its merits too)

Since this is your area I wonder if you have a preference?

Reply:
My preference is that we align the bootstrapper code to use a manifest_rev branch and detach HEAD, just as the projects.

That will solve the problem, as rev-parse then has something to look up.

Reply:
OK, great. So I will update this PR to include a patch before this one which adds a `manifest-rev` branch to the `west` repository and detaches HEAD as well. Then I can update this patch to use `{qual_manifest_rev_branch}` for the west project as well.

Reply:
Great 👍

Reply:
> But the problem was that this didn't work if the revision was a branch (which is exactly what it _must_ do and the main reason for its existence is), because the current fetching code doesn't pull in the remote's head ref space (into `refs/west/*`) unless the revision is a SHA! 😆

I think the reason for that was to minimize the need for fetching whole ref-space, but @ulfalizer knows more on the reason.